build/xtask/src/config.rs

```diff
@@ -269,35 +277,64 @@ impl Config {
         Ok(out)
     }
 
-    /// Returns a map of memory name -> range
+    /// Returns a map of memory name -> range for a specific image name
     ///
     /// This is useful when allocating memory for tasks
-    pub fn memories(&self) -> Result<IndexMap<String, Range<u32>>> {
+    pub fn memories(
+        &self,
+        image_name: &String,
+    ) -> Result<IndexMap<String, Range<u32>>> {
         self.outputs
             .iter()
             .map(|(name, out)| {
-                out.address
-                    .checked_add(out.size)
+                let region : Vec<&Output>= out.iter().filter(|o| o.name == *image_name).collect();
+                if region.len() > 1 {
+                    bail!("Multiple regions defined for image {}", image_name);
+                }
+
+                let r = region[0];
+
+                r.address
+                    .checked_add(r.size)
                     .ok_or_else(|| {
                         anyhow!(
                             "output {}: address {:08x} size {:x} would overflow",
                             name,
-                            out.address,
-                            out.size
+                            r.address,
+                            r.size
                         )
                     })
-                    .map(|end| (name.clone(), out.address..end))
+                    .map(|end| (name.clone(), r.address..end))
             })
             .collect()
     }
 
+    pub fn image_memories(
+        &self,
+        region: String,
+    ) -> Result<IndexMap<String, Range<u32>>> {
+        let mut memories: IndexMap<String, Range<u32>> = IndexMap::new();
+        for a in &self.external_images {
+            if let Some(r) = self.memories(&a)?.get(&region) {
+                memories.insert(a.clone(), r.start..r.end);
+            }
+        }
+
+        Ok(memories)
+    }
+
     /// Calculates the output region which contains the given address
     pub fn output_region(&self, vaddr: u64) -> Option<&str> {
         let vaddr = u32::try_from(vaddr).ok()?;
         self.outputs
             .iter()
             .find(|(_name, out)| {
-                vaddr >= out.address && vaddr < out.address + out.size
+                let mut stat = false;
+                for o in out.iter() {
+                    stat = stat
+                        || vaddr >= o.address && vaddr < o.address + o.size;
```

Comment:
Nit: parenthesis to clarify operator precedence (which will also cause `rustfmt` to not make this look like a closure!)

Reply:
Actually, you could replace the whole loop with something like
```
let stat = out.iter().any(|o| vaddr >= o.address && vaddr < o.address + o.size);
```

Reply:
I was hoping someone would turn this into a one liner on review (because I couldn't :))

Reply:
You could also replace `.find(..).map(..)` with `.find_map(..)`; I think clippy might suggest this? Untested, maybe typos:

```rust
self.outputs.iter().find_map(|(name, out)| {
    if vaddr >= o.address && vaddr < o.address + o.size {
        Some(name.as_str())
    } else {
        None
    }
})
```

A couple golfing ideas from this that might not be good:

1. If `Output` had a method to return the range `address..address+size`, the condition could become `o.address_range().contains(&vaddr)`
2. If going with the `.find_map()` version, the body could use `bool::then_some()` instead of an explicit if/else